File: src/test/python/workshop/function/FunctionUnitTester.py
import random

import cocotb
from cocotb.triggers import RisingEdge, Timer

from cocotblib.misc import ClockDomainAsyncReset, simulationSpeedPrinter, assertEquals, SimulationTimeout


# Send a transaction on io_cmd with a random timing
@cocotb.coroutine
def sendCmdRandomTiming(dut,payload):
    while random.random() < 0.5:
        yield RisingEdge(dut.clk)
    dut.io_cmd_valid <= 1
    dut.io_cmd_payload <= payload
    yield RisingEdge(dut.clk)
    dut.io_cmd_valid <= 0
    dut.io_cmd_payload <= 0

# Send a command with the given header + value and then check that the pin is correctly set by the hardware
@cocotb.coroutine
def driveAndCheck(dut,header,value,valueWidth,pin):
    assert(valueWidth % 8 == 0)

    for h in header:
        yield sendCmdRandomTiming(dut,ord(h))

    for i in range(int(valueWidth/8)):
        yield sendCmdRandomTiming(dut,(value >> (i*8)) & 0xFF)

    yield RisingEdge(dut.clk)
    yield RisingEdge(dut.clk)
    yield RisingEdge(dut.clk)
    yield RisingEdge(dut.clk)
    yield RisingEdge(dut.clk)
    assertEquals(pin, value,pin._name + " wasn't loaded correctly")


@cocotb.test()
def test1(dut):
    random.seed(0)

    cocotb.fork(ClockDomainAsyncReset(dut.clk, dut.reset))
    cocotb.fork(simulationSpeedPrinter(dut.clk))
    cocotb.fork(SimulationTimeout(1000*16000))

    yield driveAndCheck(dut, "setValueA",0x11,8,dut.io_valueA)
    yield driveAndCheck(dut, "setValueB", 0x22334455, 32, dut.io_valueB)
    yield driveAndCheck(dut, "setValueC", 0x66778899AABB, 48, dut.io_valueC)
    yield driveAndCheck(dut, "setValueB", 0xCAFEF00D, 32, dut.io_valueB)
    yield Timer(1000*20)

File: src/test/scala/workshop/function/FunctionUnitTester.scala
package workshop.function

import org.scalatest.FunSuite
import spinal.core._
import spinal.core.sim._
import workshop.common.WorkshopSimConfig

import scala.util.Random

//Run this scala test to generate and check that your RTL work correctly
class FunctionUnitTester extends FunSuite{
  var compiled: SimCompiled[FunctionUnit] = null

  test("compile") {
    compiled = WorkshopSimConfig().compile(FunctionUnit())
  }

  test("testbench") {
    compiled.doSim(seed = 42){dut =>
      dut.clockDomain.forkStimulus(10)
      dut.io.cmd.valid #= false

      def sendWithRandomTiming(value : BigInt) = {
        dut.clockDomain.waitSampling(Random.nextInt(4))
        dut.io.cmd.valid #= true
        dut.io.cmd.payload #= value
        dut.clockDomain.waitSampling()
        dut.io.cmd.valid #= false
        dut.io.cmd.payload.randomize()
      }

      def driveAndCheck(header : String,value : BigInt, pin : Bits) = {
        for(char <- header) {
          sendWithRandomTiming(char.toInt & 0xFF)
        }
        for(byteId <- 0 until widthOf(pin)/8) {
          sendWithRandomTiming((value >> byteId*8) & 0xFF)
        }
        dut.clockDomain.waitSampling(8)
        assert(pin.toBigInt === value, s"${pin.getName()} wasn't loaded correctly")
      }

       driveAndCheck("setValueA", 0x11l          , dut.io.valueA)
       driveAndCheck("setValueB", 0x22334455l    , dut.io.valueB)
       driveAndCheck("setValueC", 0x66778899AABBl, dut.io.valueC)
       driveAndCheck("setValueB", 0xCAFEF00Dl    , dut.io.valueB)
    }
  }
}
